DDF-1196 Improved performance of security

pklinef opened the pull request:
- The anonymous interceptor now caches and reuses the subject until
  the token is about to expire.
- Using a thread local document builder to reduce overhead of always
  creating new instances of document builders and their factories.
- Set a cache manager for abstract authorizing realm to reduce info
  logging that a cache manager was not set for all secured requests.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf-platform/47)
<!-- Reviewable:end -->

pklinef commented:
@stustison @lessarderic @jckilmer

lessarderic commented on security/filter/security-filter-login/src/main/java/org/codice/ddf/security/filter/login/LoginFilter.java:
```diff
@@ -105,12 +105,24 @@
+                        return factory.newDocumentBuilder();
+                    } catch (ParserConfigurationException ex) {
+                        throw new IllegalArgumentException("Unable to create new DocumentBuilder", ex);
+                    }
```
Shouldn't we log something here? Also, since this would leave the BUILDER null, do we need to check for null when calling BUILDER.get(), or return some kind of null object that logs and throws a SecurityServiceException or InvalidStateException?

pklinef replied:
This exception should never happen as things are configured here.

http://docs.oracle.com/javase/7/docs/api/javax/xml/parsers/ParserConfigurationException.html

lessarderic replied:
Can you add that as a comment?

lessarderic commented on security/filter/security-filter-login/src/main/java/org/codice/ddf/security/filter/login/LoginFilter.java:
```diff
@@ -253,11 +257,8 @@ private Subject handleAuthenticationToken(HttpServletRequest httpRequest, SAMLAu
-                    }
-
-                    Document doc = docBuilder.newDocument();
+                    BUILDER.get().reset();
```
Any special reason we're calling reset() now?

pklinef replied:
This is done to avoid having to create a new document builder.

"DocumentBuilder is reset to the same state as when it was created with DocumentBuilderFactory.newDocumentBuilder(). reset() is designed to allow the reuse of existing DocumentBuilders thus saving resources associated with the creation of new DocumentBuilders."

http://docs.oracle.com/javase/7/docs/api/javax/xml/parsers/DocumentBuilder.html#reset()

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -118,6 +119,22 @@
+                        factory.setNamespaceAware(false);
+                        return factory.newDocumentBuilder();
+                    } catch (ParserConfigurationException ex) {
+                        throw new IllegalArgumentException("Unable to create new DocumentBuilder", ex);
```
Same questions as above :-)

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -286,15 +303,10 @@ public void handleMessage(SoapMessage message) throws Fault {
-                                    layoutStrict = evaluateExpression(xml, xpathStrict);
-                                    layoutLaxTimestampFirst = evaluateExpression(xml, xpathLaxTimestampFirst);
-                                    layoutLaxTimestampLast = evaluateExpression(xml, xpathLaxTimestampLast);
 
```
I think those 4 variables aren't used anymore and could be removed. Unless that logic is still needed...

pklinef replied:
This code is half finished.  Eventually we will need these xpath values when we inspect the policy so I left the values for now.  I did remove the xpath evaluation for now since it was slowing down every request and then not being used.

lessarderic replied:
Can we make sure we have a ticket for this and maybe comment out the code that we need to keep for it and put TODO - Ticket #.

pklinef replied:
I added a TODO with ticket number at the top of this big if else block

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -286,15 +303,10 @@ public void handleMessage(SoapMessage message) throws Fault {
                                 } else if (qName.equals(SP12Constants.TRANSPORT_TOKEN)) {
                                 } else if (qName.equals(SP12Constants.HTTPS_TOKEN)) {
                                     String xpath = "/HttpsToken/Policy/RequireClientCertificate";
-                                    requireClientCert = evaluateExpression(xml, xpath);
```
Looks like requireClientCert isn't needed anymore either and this else if block could be empty.

pklinef replied:
This will be added in the future.  I left these in after talking to @stustison.

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -376,18 +388,23 @@ public void handleMessage(SoapMessage message) throws Fault {
+            synchronized (lock) {
+                try {
+                    subject = securityManager.getSubject(token);
+                    anonymousSubject = subject;
```
Not clear why we need the temporary subject variable here.

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -376,18 +388,23 @@ public void handleMessage(SoapMessage message) throws Fault {
-                subject = securityManager.getSubject(token);
-            } catch (SecurityServiceException sse) {
-                LOGGER.warn("Unable to request subject for anonymous user.", sse);
+        Subject subject = anonymousSubject;
```
Instead of growing this already big class, shouldn't we move the token management piece to a separate, thread-safe class?

pklinef replied:
I am not sure how reusable the code would be outside of this class.  I can extract the subject logic to make this function more readable.

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -376,18 +388,23 @@ public void handleMessage(SoapMessage message) throws Fault {
+        } else {
+            LOGGER.debug("Using cached Anonymous user token");
         }
+
```
The only way we can get inside the following if statement is if the securityManager.getSubject() returns null, right? Looking at the SecurityManagerImpl class, I'm not sure this is possible.

pklinef replied:
If anonymousSubject is null and securityManager.getSubject throws an exception, then it is possible for the value to still be null.  This could happen for example if there is a connection issue with the STS.

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -413,6 +430,13 @@ private void createSecurityToken(SoapVersion version, SOAPFactory soapFactory, S
         }
     }
 
+    private boolean tokenAboutToExpire() {
```
Nice!

lessarderic commented on security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java:
```diff
@@ -553,15 +575,10 @@ private boolean evaluateExpression(String xml, String xpathStmt) {
-        dbf.setNamespaceAware(false);
-        DocumentBuilder db = null;
         Document document = null;
+        BUILDER.get().reset();
```
Ok, now I think I know why we need the reset() above :-)

lessarderic commented:
Where are the unit tests?

pklinef commented:
This code is tested by the integration tests.  It is hard to unit test this code given how coupled it is to CXF.

pklinef commented:
@mverret

stustison commented:
:+1:

lessarderic commented:
:+1:
Let me know if you make any more changes and want me to take another look.

jckilmer commented:
Review status: all files reviewed, 9 unresolved discussions, some commit checks broke.
Reviewed files:
- security/core/security-core-impl/src/main/java/ddf/security/service/impl/AbstractAuthorizingRealm.java @ r1
- security/filter/security-filter-login/src/main/java/org/codice/ddf/security/filter/login/LoginFilter.java @ r1
- security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java @ r1

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-platform/47)
<!-- Sent from Reviewable.io -->

pklinef commented:
Review status: 1 of 3 files reviewed, 6 unresolved discussions, some commit checks pending.

---

<sup>**[security/filter/security-filter-login/src/main/java/org/codice/ddf/security/filter/login/LoginFilter.java, line 122 \[r1\]](https://reviewable.io:443/reviews/codice/ddf-platform/47#-JpdClIIiYScQVlw2qK6-r1-122)** ([raw file](https://github.com/codice/ddf-platform/blob/8ee9f024709840147e8fdb56c7bb1322a5bd2478/security/filter/security-filter-login/src/main/java/org/codice/ddf/security/filter/login/LoginFilter.java#L122)):</sup>
I added a comment for both files.

---

<sup>**[security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java, line 399 \[r1\]](https://reviewable.io:443/reviews/codice/ddf-platform/47#-JpdClIIiYScQVlw2qK7-r1-399)** ([raw file](https://github.com/codice/ddf-platform/blob/8ee9f024709840147e8fdb56c7bb1322a5bd2478/security/interceptor/security-interceptor-anonymous/src/main/java/org/codice/ddf/security/interceptor/AnonymousInterceptor.java#L399)):</sup>
Good point.  I think I did this while I was experimenting and forgot to clean it up.  Fixed.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-platform/47)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 2 of 4 files reviewed, all discussions resolved, some commit checks broke.
Reviewed files:
- security/interceptor/security-interceptor-anonymous/src/main/resources/OSGI-INF/blueprint/blueprint.xml @ r2

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf-platform/47)
<!-- Sent from Reviewable.io -->
